File: ios/MoWalk/Services/StepDeltaCalculator.swift
import Foundation

final class StepDeltaCalculator {
    private var bootOffset: Int?
    private var previousDayTotal: Int = 0
    private var lastKnownToday: Int = 0

    func setYesterdayTotal(_ total: Int) {
        previousDayTotal = total
    }

    func update(cumulativeSteps: Int, yesterdayTotal: Int) -> Int {
        previousDayTotal = yesterdayTotal

        guard let offset = bootOffset else {
            bootOffset = cumulativeSteps
            return 0
        }

        let rawToday = cumulativeSteps - offset - yesterdayTotal
        let todaySteps = max(rawToday, lastKnownToday)
        lastKnownToday = todaySteps
        return todaySteps
    }

    func onDeviceReboot(cumulativeSteps: Int, yesterdayTotal: Int) -> Int {
        bootOffset = cumulativeSteps
        previousDayTotal = yesterdayTotal
        lastKnownToday = 0
        return 0
    }

    func resetForNewDay(yesterdayTotal: Int) {
        bootOffset = nil
        previousDayTotal = yesterdayTotal
        lastKnownToday = 0
    }

    func getLastKnownToday() -> Int { lastKnownToday }
    func getYesterdayTotal() -> Int { previousDayTotal }
}

File: android/app/src/main/java/com/mowalk/app/data/sensor/StepDeltaCalculator.kt
package com.mowalk.app.data.sensor

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class StepDeltaCalculator {

    private val _bootOffset = MutableStateFlow<Int?>(null)
    private val _previousDayTotal = MutableStateFlow(0)
    private val _lastKnownToday = MutableStateFlow(0)

    fun setYesterdayTotal(total: Int) {
        _previousDayTotal.value = total
    }

    fun update(cumulativeSteps: Int, yesterdayTotal: Int): Int {
        _previousDayTotal.value = yesterdayTotal

        val bootOffset = _bootOffset.value
        if (bootOffset == null) {
            _bootOffset.value = cumulativeSteps
            return 0
        }

        val rawToday = cumulativeSteps - bootOffset - yesterdayTotal
        val todaySteps = maxOf(rawToday, _lastKnownToday.value)
        _lastKnownToday.value = todaySteps
        return todaySteps
    }

    fun onDeviceReboot(cumulativeSteps: Int, yesterdayTotal: Int): Int {
        _bootOffset.value = cumulativeSteps
        _previousDayTotal.value = yesterdayTotal
        _lastKnownToday.value = 0
        return 0
    }

    fun resetForNewDay(yesterdayTotal: Int) {
        _bootOffset.value = null
        _previousDayTotal.value = yesterdayTotal
        _lastKnownToday.value = 0
    }

    fun getLastKnownToday(): Int = _lastKnownToday.value
    fun getYesterdayTotal(): Int = _previousDayTotal.value
}
